lib/scope.go

```diff
@@ -44,54 +47,65 @@ func (s *scope) ActiveProfile() *profile.Profile {
 	return s.pro
 }
 
+// Bus returns the event bus
+func (s *scope) Bus() event.Bus {
+	// TODO(dustmop): Filter only events for this scope.
+	return s.inst.bus
+}
+
+// Config returns the config
+func (s *scope) Config() *config.Config {
+	return s.inst.cfg
+}
+
 // Context returns the context for this scope. Though this pattern is usually discouraged,
 // we're following http.Request's lead, as scope plays the same role. The lifetime of a
 // single scope matches the lifetime of the Context; this ownership is not long-lived
 func (s *scope) Context() context.Context {
 	return s.ctx
 }
 
+// Dscache returns the dscache
+func (s *scope) Dscache() *dscache.Dscache {
+	return s.inst.Dscache()
+}
+
 // FSISubsystem returns a reference to the FSI subsystem
 // TODO(dustmop): This subsystem contains global data, we should move that data out and
 // into scope
 func (s *scope) FSISubsystem() *fsi.FSI {
 	return s.inst.fsi
 }
 
-// Bus returns the event bus
-func (s *scope) Bus() event.Bus {
-	// TODO(dustmop): Filter only events for this scope.
-	return s.inst.bus
-}
-
 // Filesystem returns a filesystem
 func (s *scope) Filesystem() *muxfs.Mux {
 	return s.inst.qfs
 }
 
-// Config returns the config
-func (s *scope) Config() *config.Config {
-	return s.inst.cfg
+// GetVersionInfoShim is in the process of being removed. Try not to add new callers.
+func (s *scope) GetVersionInfoShim(ref dsref.Ref) (*dsref.VersionInfo, error) {
+	r := s.inst.Repo()
+	return repo.GetVersionInfoShim(r, ref)
 }
 
-// Dscache returns the dscache
-func (s *scope) Dscache() *dscache.Dscache {
-	return s.inst.Dscache()
+// Loader returns the instance
+func (s *scope) Loader() *Instance {
```

Comment:
Loader's return value should be `dsref.Loader`, just the interface that is able to load datasets. Returning the instance itself go against the desire to migrate state management over to scope.

For situations where `Instance` is needed for functionality to work, add a new function to `scope` with the same parameters and have that function call the `instance` version. This isolates the exposure of `instance`'s state and let's us know exactly where it is being used.

For example, I know `lib/config.go` is using `scope.Loader().ChangeConfig(...)`. Instead, `ChangeConfig` should be added to `scope` and that should be called instead.